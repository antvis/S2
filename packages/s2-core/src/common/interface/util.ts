export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
