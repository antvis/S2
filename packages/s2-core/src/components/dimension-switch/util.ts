import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
