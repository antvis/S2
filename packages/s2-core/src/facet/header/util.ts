import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
