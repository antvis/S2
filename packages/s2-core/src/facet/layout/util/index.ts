export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
