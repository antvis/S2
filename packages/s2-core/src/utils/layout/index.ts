export { generateId } from './generate-id';
export * from './transform-custom-tree-items';
export * from './frozen';
