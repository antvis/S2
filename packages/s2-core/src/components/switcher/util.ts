import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
