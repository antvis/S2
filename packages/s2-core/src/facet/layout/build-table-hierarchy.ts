import { map } from 'lodash';
import { TableHeaderParams } from '@/facet/layout/interface';
import { SERIES_NUMBER_FIELD } from '@/common/constant';
import { i18n } from '@/common/i18n';
import { generateHeaderNodes } from '@/utils/layout/generate-header-nodes';

export const buildTableHierarchy = (params: TableHeaderParams) => {
  const { facetCfg, hierarchy, parentNode } = params;
  const { columns, spreadsheet, dataSet } = facetCfg;

  const hiddenColumnsDetail = spreadsheet.store.get('hiddenColumnsDetail');
  const showSeriesNumber = facetCfg?.showSeriesNumber;

  const displayedColumns = columns.filter((column) => {
    if (!hiddenColumnsDetail) {
      return true;
    }
    return hiddenColumnsDetail.every((detail) => {
      return detail.hideColumnNodes.every((node) => {
        return node.field !== column;
      });
    });
  });
  const fields = [...displayedColumns];

  const fieldValues = map(displayedColumns, (val) => dataSet.getFieldName(val));

  if (showSeriesNumber) {
    fields.unshift(SERIES_NUMBER_FIELD);
    fieldValues.unshift(i18n('序号'));
  }

  generateHeaderNodes({
    currentField: displayedColumns[0],
    fields,
    fieldValues,
    facetCfg,
    hierarchy,
    parentNode,
    level: 0,
    query: {},
    addMeasureInTotalQuery: false,
    addTotalMeasureInTotal: false,
  });
};
