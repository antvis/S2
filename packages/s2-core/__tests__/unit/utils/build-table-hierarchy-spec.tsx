import { act } from 'react-dom/test-utils';
import 'antd/dist/antd.min.css';
import ReactDOM from 'react-dom';
import React from 'react';
import { getContainer, getMockData } from '../../util/helpers';
import {
  S2DataConfig,
  S2Options,
  SheetComponent,
  SpreadSheet,
  PivotSheet,
  Node,
  Hierarchy,
} from '@/index';
import { buildTableHierarchy } from '@/facet/layout/build-table-hierarchy';

const data = getMockData('../data/tableau-supermarket.csv');

let spreadsheetIns: SpreadSheet;

const getSpreadSheet = (
  dom: string | HTMLElement,
  dataCfg: S2DataConfig,
  options: S2Options,
) => {
  spreadsheetIns = new PivotSheet(dom, dataCfg, options);
  return spreadsheetIns;
};

const getDataCfg = () => {
  return {
    fields: {
      columns: [
        'area',
        'province',
        'city',
        'type',
        'sub_type',
        'profit',
        'count',
      ],
    },
    data,
  };
};

const getOptions = () => {
  return {
    width: 800,
    height: 600,
    showSeriesNumber: true,
    mode: 'table',
    style: {
      colCfg: {
        colWidthType: 'compact',
      },
      cellCfg: {
        height: 32,
      },
      device: 'pc',
    },
  };
};

function MainLayout(props) {
  return (
    <div>
      <div style={{ display: 'inline-block' }}></div>
      <SheetComponent
        dataCfg={props.dataCfg}
        adaptive={false}
        options={props.options}
        spreadsheet={getSpreadSheet}
      />
    </div>
  );
}

describe('buildTableHierarchy', () => {
  test('should generate nodes and hierarchy correctly', () => {
    act(() => {
      ReactDOM.render(
        <MainLayout dataCfg={getDataCfg()} options={getOptions()} />,
        getContainer(),
      );
    });

    const rootNode = Node.rootNode();
    const hierarchy = new Hierarchy();

    buildTableHierarchy({
      parentNode: rootNode,
      facetCfg: spreadsheetIns.facet.cfg,
      hierarchy,
    });

    expect(rootNode).toMatchSnapshot();
    expect(hierarchy).toMatchSnapshot();
  });
});
