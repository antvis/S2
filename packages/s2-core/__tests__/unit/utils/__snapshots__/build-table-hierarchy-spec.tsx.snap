// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTableHierarchy should generate nodes and hierarchy correctly 1`] = `
Object {
  "belongsCell": undefined,
  "children": Array [
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]序号",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "序号",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "序号",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "序号",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]area",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "area",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "area",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "area",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]province",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "province",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "province",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "province",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]city",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "city",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "city",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "city",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]type",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "type",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "type",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "type",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]sub_type",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "sub_type",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "sub_type",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "sub_type",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]profit",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "profit",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "profit",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "profit",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]count",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "count",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "count",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "count",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]数值",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "数值",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "数值",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "数值",
      "width": 0,
      "x": 0,
      "y": 0,
    },
  ],
  "colIndex": -1,
  "field": undefined,
  "height": 0,
  "id": "root",
  "inCollapseNode": undefined,
  "isCollapsed": undefined,
  "isGrandTotals": undefined,
  "isLeaf": undefined,
  "isPivotMode": undefined,
  "isSubTotals": undefined,
  "isTotalMeasure": undefined,
  "isTotals": undefined,
  "key": "",
  "label": "",
  "level": undefined,
  "originalHeight": 0,
  "originalWidth": 0,
  "padding": 0,
  "query": undefined,
  "rowIndex": undefined,
  "seriesNumberWidth": undefined,
  "toJSON": [Function],
  "value": "",
  "width": 0,
  "x": 0,
  "y": 0,
}
`;

exports[`buildTableHierarchy should generate nodes and hierarchy correctly 2`] = `
Hierarchy {
  "allNodesWithoutRoot": Array [
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]序号",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "序号",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "序号",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "序号",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]area",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "area",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "area",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "area",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]province",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "province",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "province",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "province",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]city",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "city",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "city",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "city",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]type",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "type",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "type",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "type",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]sub_type",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "sub_type",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "sub_type",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "sub_type",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]profit",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "profit",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "profit",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "profit",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]count",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "count",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "count",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "count",
      "width": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]数值",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "数值",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "数值",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "数值",
      "width": 0,
      "x": 0,
      "y": 0,
    },
  ],
  "height": 0,
  "indexNode": Array [],
  "maxLevel": 0,
  "sampleNodeForLastLevel": Object {
    "belongsCell": undefined,
    "children": Array [],
    "colIndex": -1,
    "field": "area",
    "height": 0,
    "id": "root[&]序号",
    "inCollapseNode": undefined,
    "isCollapsed": false,
    "isGrandTotals": false,
    "isLeaf": false,
    "isPivotMode": undefined,
    "isSubTotals": false,
    "isTotalMeasure": false,
    "isTotals": false,
    "key": "area",
    "label": "序号",
    "level": 0,
    "originalHeight": 0,
    "originalWidth": 0,
    "padding": 0,
    "query": Object {
      "area": "序号",
    },
    "rowIndex": undefined,
    "seriesNumberWidth": undefined,
    "toJSON": [Function],
    "value": "序号",
    "width": 0,
    "x": 0,
    "y": 0,
  },
  "sampleNodesForAllLevels": Array [
    Object {
      "belongsCell": undefined,
      "children": Array [],
      "colIndex": -1,
      "field": "area",
      "height": 0,
      "id": "root[&]序号",
      "inCollapseNode": undefined,
      "isCollapsed": false,
      "isGrandTotals": false,
      "isLeaf": false,
      "isPivotMode": undefined,
      "isSubTotals": false,
      "isTotalMeasure": false,
      "isTotals": false,
      "key": "area",
      "label": "序号",
      "level": 0,
      "originalHeight": 0,
      "originalWidth": 0,
      "padding": 0,
      "query": Object {
        "area": "序号",
      },
      "rowIndex": undefined,
      "seriesNumberWidth": undefined,
      "toJSON": [Function],
      "value": "序号",
      "width": 0,
      "x": 0,
      "y": 0,
    },
  ],
  "width": 0,
}
`;
